### PR TITLE
Add docs to explain inbound/outbound

### LINKED ISSUE #138: Write narrative documentation to explain what Inbound/Outbound mean
Given that `channelRead` is the most prominent `ChannelInboundHandler` method, and that `write` is the most prominent `ChannelOutboundHandler` method, it's easy for users to assume that inbound means reads, and outbound means writes. This will cause user confusion when they discover that `channelWritabilityChanged` is an Inbound method, and that `read` is an Outbound one!

We should write some narrative docs for these two protocols to make it clear that this is not actually how they work: instead, that inbound/outbound is about *causality*. Specifically, Inbound events are things the socket tells the pipeline, while Outbound events are things the pipeline tells the socket. From that construction it becomes clear why `read` is Outbound (the pipeline tells the socket to do a read), and why `channelWritabilityChanged` is Inbound (the socket tells the pipeline whether it's writable or not).

### PR DESCRIPTION
**Pull request by Lukasa**
Resolves #138.

### CODE REVIEW

File: Sources/NIO/ChannelHandler.swift [hunk 1]
```diff
@@ -33,7 +33,12 @@ public protocol ChannelHandler: class {
 
 /// Untyped `ChannelHandler` which handles outbound I/O events or intercept an outbound I/O operation.
 ///
-/// We _strongly_ advice against implementing this protocol directly. Please implement `ChannelOutboundHandler`.
+/// Despite the fact that `write` is one of the methods on this `protocol`, you should avoid assuming that "outbound" events are to do with
+/// writing to sockets. Instead, "outbound" events are events that are passed *to* the socket (or other event source): that is, things you tell
+/// the socket to do. That includes `write` ("write this data to the socket"), but it also includes `read` ("please begin attempting to read from
+/// the socket") and `bind` ("please bind the following address"), which have nothing to do with sending data.
```
Comment (normanmaurer) [inline Sources/NIO/ChannelHandler.swift]: @Lukasa I am not sure we should use `socket` here as a `Channel` may not be backed by a `socket`.

File: Sources/NIO/ChannelHandler.swift [hunk 2]
```diff
@@ -124,7 +129,12 @@ public protocol _ChannelOutboundHandler: ChannelHandler {
 
 /// Untyped `ChannelHandler` which handles inbound I/O events.
 ///
-/// We _strongly_ advice against implementing this protocol directly. Please implement `ChannelInboundHandler`.
+/// Despite the fact that `channelRead` is one of the methods on this `protocol`, you should avoid assuming that "inbound" events are to do with
+/// reading from sockets. Instead, "inbound" events are events that originate *from* the socket (or other event source): that is, events that the
```
Comment (normanmaurer) [inline Sources/NIO/ChannelHandler.swift]: same comment as above...

File: Sources/NIO/ChannelHandler.swift (hunk 1 above)
Comment (normanmaurer) [reply Sources/NIO/ChannelHandler.swift]: That said the rest of the docs sounds good... Maybe just make it clear that `socket` is an example here.

**Comment by Lukasa** [comment]
Ok, addressed.

**Comment by helje5** [comment]
It was me which was hitting this, we talked about it in the Slack. I think those comments are buried too deep and this somehow belongs into the main README or some other document. (Now that I think I understood it, I consider writing a blog entry about it.)

What Inbound and Outbound is, is not obvious to someone who didn't do Netty before (actually I think the naming in general is not particularly good, no offence!)

I (and I suspect many will) originally mapped "Inbound" to the concept of "ReadStream" and "Outbound" to the concept of a "WriteStream". But this is not at all what is going on here. I'm not yet sure how to describe it easily, but the way I understand it Outbound is like a "perform action" and Inbound is like "did perform action". It decouples the emitter from the result handler (in a pretty "distant" way).
(and a thing which makes this harder to understand is `autoRead` :-) )

**Comment by normanmaurer** [comment]
@helje5 I am happy for other naming suggestions. We used `ChannelUpstreamHandler` and `ChannelDownstreamHandler` in netty 3 which was even more confusing to people. Honestly most people in netty land like the inbound / outbound naming, but maybe its just because the old names were even worse 🤣

**Comment by helje5** [comment]
You are right, I shouldn't complain w/o providing decent alternatives ;->

I'm not sure yet, I don't fully understand yet _why_ you do some things the way you do them. Maybe I come back to you on this later.

For this particular issue, it would be cool to get some more prominent mention of this on the main README or design document (actually I recommend making the README smaller, and put the current stuff in a DESIGN.md or sth).

**Comment by weissi** [comment]
@normanmaurer happy with this?

**Comment by normanmaurer** [comment]
@weissi yep... missed that @Lukasa did address my comment. Let me merge FTW!

**Pull request merged**